File: typescript/sdk/src/uri.ts
/**
 * URI parsing and building utilities for skill:// URIs.
 *
 * Supports multi-segment skill paths per the Skills Extension SEP:
 *   - skill://code-review/SKILL.md              (single-segment)
 *   - skill://acme/billing/refunds/SKILL.md     (multi-segment)
 *   - skill://acme/billing/refunds/templates/email.md  (supporting file)
 *
 * Per the SEP: the final segment of <skill-path> MUST equal the skill's
 * frontmatter name. Preceding segments are a server-chosen organizational
 * prefix. In skill://acme/billing/refunds/SKILL.md, the prefix is
 * "acme/billing" and the skill name is "refunds".
 */

/** The skill:// URI scheme prefix. */
export const SKILL_URI_SCHEME = "skill://";

/** Default skill content filename. */
export const SKILL_FILENAME = "SKILL.md";

/** Well-known URI for the skill index (SEP discovery mechanism). */
export const INDEX_JSON_URI = "skill://index.json";

/**
 * Agent Skills naming rule: skill names contain only lowercase letters,
 * digits, and hyphens. Per SEP-2640, the final segment of `<skill-path>` —
 * which equals the frontmatter `name` — MUST satisfy this rule. The rule
 * also underpins the SEP's reservation note that `index.json` cannot
 * collide with a skill name.
 */
const SKILL_NAME_REGEX = /^[a-z0-9-]+$/;

/**
 * Check whether a string satisfies the Agent Skills `name` field rule:
 * lowercase letters, digits, and hyphens, non-empty.
 */
export function isValidSkillName(name: string): boolean {
  return SKILL_NAME_REGEX.test(name);
}

/**
 * Parsed components of a skill:// URI.
 */
export interface ParsedSkillUri {
  /** Multi-segment skill path (e.g., "acme/billing/refunds") */
  skillPath: string;
  /** File path within the skill (e.g., "SKILL.md", "templates/email.md") */
  filePath: string;
}

/**
 * Parse a skill:// URI into skill path and file path components.
 *
 * For SKILL.md URIs, the split is unambiguous because the last segment is
 * a known sentinel. For supporting file URIs, the caller must use
 * resolveSkillFileUri() with known skill paths.
 *
 * Returns null if the URI doesn't match the skill:// scheme or is the
 * special index.json URI.
 *
 * Examples:
 *   "skill://code-review/SKILL.md"
 *     → { skillPath: "code-review", filePath: "SKILL.md" }
 *   "skill://acme/billing/refunds/SKILL.md"
 *     → { skillPath: "acme/billing/refunds", filePath: "SKILL.md" }
 */
export function parseSkillUri(uri: string): ParsedSkillUri | null {
  if (!uri.startsWith(SKILL_URI_SCHEME)) return null;

  const rest = uri.slice(SKILL_URI_SCHEME.length);
  if (!rest || rest === "index.json") return null;

  const slashIndex = rest.lastIndexOf("/");
  if (slashIndex === -1) return null;

  const beforeLast = rest.slice(0, slashIndex);
  const afterLast = rest.slice(slashIndex + 1);

  // Known sentinel: SKILL.md as the last segment
  if (afterLast === SKILL_FILENAME || afterLast.toLowerCase() === "skill.md") {
    return { skillPath: beforeLast, filePath: afterLast };
  }

  // For arbitrary file paths, we can't determine the split from the URI alone.
  // Return with empty skillPath — caller should use resolveSkillFileUri().
  return { skillPath: "", filePath: rest };
}

/**
 * Resolve a skill:// URI for a supporting file by matching against known
 * skill paths. Uses longest-prefix matching to handle nested hierarchies.
 *
 * Example:
 *   resolveSkillFileUri(
 *     "skill://acme/billing/refunds/templates/email.md",
 *     ["code-review", "acme/billing/refunds", "acme/onboarding"]
 *   )
 *   → { skillPath: "acme/billing/refunds", filePath: "templates/email.md" }
 */
export function resolveSkillFileUri(
  uri: string,
  knownSkillPaths: string[],
): ParsedSkillUri | null {
  if (!uri.startsWith(SKILL_URI_SCHEME)) return null;

  const rest = uri.slice(SKILL_URI_SCHEME.length);

  // Sort by length descending for longest-prefix match
  const sorted = [...knownSkillPaths].sort((a, b) => b.length - a.length);
  for (const sp of sorted) {
    if (rest.startsWith(sp + "/")) {
      return { skillPath: sp, filePath: rest.slice(sp.length + 1) };
    }
  }

  return null;
}

/**
 * Build a skill:// URI from a multi-segment skill path and optional file path.
 * Defaults to SKILL.md if no file path is provided.
 *
 * Examples:
 *   buildSkillUri("acme/billing/refunds")
 *     → "skill://acme/billing/refunds/SKILL.md"
 *   buildSkillUri("code-review", "references/REFERENCE.md")
 *     → "skill://code-review/references/REFERENCE.md"
 */
export function buildSkillUri(skillPath: string, filePath?: string): string {
  return `${SKILL_URI_SCHEME}${skillPath}/${filePath ?? SKILL_FILENAME}`;
}

/**
 * Check if a URI points to a skill's SKILL.md content.
 */
export function isSkillContentUri(uri: string): boolean {
  const parsed = parseSkillUri(uri);
  return (
    parsed !== null &&
    (parsed.filePath === SKILL_FILENAME ||
      parsed.filePath.toLowerCase() === "skill.md")
  );
}

/**
 * Check if a URI is the well-known skill index resource.
 */
export function isIndexJsonUri(uri: string): boolean {
  return uri === INDEX_JSON_URI;
}

/**
 * Extract the `<skill-path>` from any-scheme URI ending in `/SKILL.md`.
 *
 * Per SEP-2640, the structural constraints on `<skill-path>` (final segment
 * equals the skill name, `SKILL.md` explicit, no nesting) apply regardless
 * of scheme. So for `github://owner/repo/skills/refunds/SKILL.md` the
 * skill-path is `owner/repo/skills/refunds`.
 *
 * Returns null if the URI doesn't have the form `<scheme>://<path>/SKILL.md`
 * (case-insensitive on the filename).
 */
export function extractSkillPathFromUri(uri: string): string | null {
  const schemeMatch = uri.match(/^[a-zA-Z][a-zA-Z0-9+.-]*:\/\/(.*)$/);
  if (!schemeMatch) return null;

  const rest = schemeMatch[1];
  const slashIndex = rest.lastIndexOf("/");
  if (slashIndex <= 0) return null;

  const lastSegment = rest.slice(slashIndex + 1);
  if (lastSegment !== SKILL_FILENAME && lastSegment.toLowerCase() !== "skill.md") {
    return null;
  }

  const skillPath = rest.slice(0, slashIndex);
  return skillPath || null;
}

File: python/sdk/src/mcp_experimental_ext_skills/_uri.py
"""URI parsing and building utilities for ``skill://`` URIs.

Supports multi-segment skill paths per SEP-2640::

    skill://code-review/SKILL.md                       (single-segment)
    skill://acme/billing/refunds/SKILL.md              (multi-segment)
    skill://acme/billing/refunds/templates/email.md    (supporting file)

Per SEP-2640: the final segment of ``<skill-path>`` MUST equal the skill's
frontmatter name. Preceding segments are a server-chosen organizational
prefix.

**Important**: per SEP-2640 §Resource Mapping, "the first segment of
``<skill-path>`` occupies the authority component" but "carries no special
semantics under this convention and clients MUST NOT attempt DNS or
network resolution of it." This means we **cannot** use
:func:`urllib.parse.urlparse` here — it would treat the first segment as a
host and split the URI incorrectly. We use raw string slicing, mirroring
the TS SDK's ``parseSkillUri`` implementation exactly.
"""

from __future__ import annotations

from dataclasses import dataclass

#: The ``skill://`` URI scheme prefix.
SKILL_URI_SCHEME = "skill://"

#: Default skill content filename.
SKILL_FILENAME = "SKILL.md"

#: Well-known URI for the skill index (SEP discovery mechanism).
INDEX_JSON_URI = "skill://index.json"


@dataclass(frozen=True)
class ParsedSkillUri:
    """Parsed components of a ``skill://`` URI."""

    skill_path: str
    """Multi-segment skill path (e.g., "acme/billing/refunds")."""

    file_path: str
    """File path within the skill (e.g., "SKILL.md", "templates/email.md")."""


def parse_skill_uri(uri: str) -> ParsedSkillUri | None:
    """Parse a ``skill://`` URI into skill path and file path components.

    For SKILL.md URIs, the split is unambiguous because the last segment is
    a known sentinel. For supporting file URIs, the caller must use
    :func:`resolve_skill_file_uri` with known skill paths.

    Returns ``None`` if the URI doesn't match the ``skill://`` scheme or is
    the special index.json URI.

    Examples::

        parse_skill_uri("skill://code-review/SKILL.md")
        # -> ParsedSkillUri(skill_path="code-review", file_path="SKILL.md")

        parse_skill_uri("skill://acme/billing/refunds/SKILL.md")
        # -> ParsedSkillUri(skill_path="acme/billing/refunds", file_path="SKILL.md")
    """
    if not uri.startswith(SKILL_URI_SCHEME):
        return None

    rest = uri[len(SKILL_URI_SCHEME) :]
    if not rest or rest == "index.json":
        return None

    slash_index = rest.rfind("/")
    if slash_index == -1:
        return None

    before_last = rest[:slash_index]
    after_last = rest[slash_index + 1 :]

    # Per SEP-2640 §Skill Format, the file is spelled ``SKILL.md``
    # (uppercase). We match exactly to keep URI generation and parsing
    # symmetric across implementations.
    if after_last == SKILL_FILENAME:
        return ParsedSkillUri(skill_path=before_last, file_path=after_last)

    # For arbitrary file paths, we can't determine the split from the URI
    # alone. Return with empty skill_path — caller should use
    # resolve_skill_file_uri().
    return ParsedSkillUri(skill_path="", file_path=rest)


def resolve_skill_file_uri(
    uri: str,
    known_skill_paths: list[str],
) -> ParsedSkillUri | None:
    """Resolve a ``skill://`` URI for a supporting file by matching against
    known skill paths.

    Uses longest-prefix matching to handle nested hierarchies.

    Example::

        resolve_skill_file_uri(
            "skill://acme/billing/refunds/templates/email.md",
            ["code-review", "acme/billing/refunds", "acme/onboarding"],
        )
        # -> ParsedSkillUri(skill_path="acme/billing/refunds",
        #                   file_path="templates/email.md")
    """
    if not uri.startswith(SKILL_URI_SCHEME):
        return None

    rest = uri[len(SKILL_URI_SCHEME) :]

    # Sort by length descending for longest-prefix match
    sorted_paths = sorted(known_skill_paths, key=len, reverse=True)
    for sp in sorted_paths:
        if rest.startswith(sp + "/"):
            return ParsedSkillUri(skill_path=sp, file_path=rest[len(sp) + 1 :])

    return None


def build_skill_uri(skill_path: str, file_path: str | None = None) -> str:
    """Build a ``skill://`` URI from a multi-segment skill path and optional
    file path.

    Defaults to SKILL.md if no file path is provided.

    Examples::

        build_skill_uri("acme/billing/refunds")
        # -> "skill://acme/billing/refunds/SKILL.md"

        build_skill_uri("code-review", "references/REFERENCE.md")
        # -> "skill://code-review/references/REFERENCE.md"
    """
    return f"{SKILL_URI_SCHEME}{skill_path}/{file_path or SKILL_FILENAME}"


def is_skill_content_uri(uri: str) -> bool:
    """Check if a URI points to a skill's SKILL.md content.

    Per SEP-2640 §Skill Format, the file is spelled ``SKILL.md`` (uppercase).
    """
    parsed = parse_skill_uri(uri)
    if parsed is None:
        return False
    return parsed.file_path == SKILL_FILENAME


def is_index_json_uri(uri: str) -> bool:
    """Check if a URI is the well-known skill index resource."""
    return uri == INDEX_JSON_URI


__all__ = [
    "INDEX_JSON_URI",
    "SKILL_FILENAME",
    "SKILL_URI_SCHEME",
    "ParsedSkillUri",
    "build_skill_uri",
    "is_index_json_uri",
    "is_skill_content_uri",
    "parse_skill_uri",
    "resolve_skill_file_uri",
]
